lightning/checkpoints/checkpoints.go

```diff
@@ -329,8 +332,60 @@ type DestroyedTableCheckpoint struct {
 	MaxEngineID int32
 }
 
+type TaskCheckpoint struct {
+	TaskId       int64
+	SourceDir    string
+	Backend      string
+	ImporterAddr string
+	TiDBHost     string
+	TiDBPort     int
+	PdAddr       string
+	SortedKVDir  string
+}
+
+func verifyCheckpoint(cfg *config.Config, taskCp *TaskCheckpoint) error {
+	if taskCp == nil {
+		return nil
+	}
+	// always check the backend value even with 'check-requirements = false'
+	retryUsage := "destroy all checkpoints and remove all restored tables and try again"
+	if cfg.TikvImporter.Backend != taskCp.Backend {
+		return errors.Errorf("config 'tikv-importer.backend' value '%s' different from checkpoint value '%s', please %s", cfg.TikvImporter.Backend, taskCp.Backend, retryUsage)
+	}
+
+	if cfg.App.CheckRequirements {
+		errorFmt := "config '%s' value '%s' different from checkpoint value '%s'. You may set 'check-requirements = false' to skip this check or " + retryUsage
+		if cfg.Mydumper.SourceDir != taskCp.SourceDir {
+			return errors.Errorf(errorFmt, "mydumper.data-source-dir", cfg.Mydumper.SourceDir, taskCp.SourceDir)
+		}
+
+		if cfg.TikvImporter.Backend == config.BackendLocal && cfg.TikvImporter.SortedKVDir != taskCp.SortedKVDir {
+			return errors.Errorf(errorFmt, "mydumper.sorted-kv-dir", cfg.TikvImporter.SortedKVDir, taskCp.SortedKVDir)
+		}
+
+		if cfg.TikvImporter.Backend == config.BackendImporter && cfg.TikvImporter.Addr != taskCp.ImporterAddr {
+			return errors.Errorf(errorFmt, "tikv-importer.addr", cfg.TikvImporter.Backend, taskCp.Backend)
+		}
+
+		if cfg.TiDB.Host != taskCp.TiDBHost {
+			return errors.Errorf(errorFmt, "tidb.host", cfg.TiDB.Host, taskCp.TiDBHost)
+		}
+
+		if cfg.TiDB.Port != taskCp.TiDBPort {
+			return errors.Errorf(errorFmt, "tidb.port", cfg.TiDB.Port, taskCp.TiDBPort)
+		}
+
+		if cfg.TiDB.PdAddr != taskCp.PdAddr {
+			return errors.Errorf(errorFmt, "tidb.pd-addr", cfg.TiDB.PdAddr, taskCp.PdAddr)
+		}
+	}
+
+	return nil
+}
+
 type CheckpointsDB interface {
-	Initialize(ctx context.Context, dbInfo map[string]*TidbDBInfo) error
+	Initialize(ctx context.Context, cfg *config.Config, dbInfo map[string]*TidbDBInfo) error
+	getTaskCheckpoint(ctx context.Context) (*TaskCheckpoint, error)
```

Comment:
i suggest we either change the interface to 

```suggestion
	Initialize(ctx context.Context, dbInfo map[string]*TidbDBInfo) (*TaskCheckpoint, error)
```

or make `getTaskCheckpoint` public so it is reachable from `restore.go`

```suggestion
	Initialize(ctx context.Context, dbInfo map[string]*TidbDBInfo) error
 	TaskCheckpoint(ctx context.Context) (*TaskCheckpoint, error)
```

in any case, perform `verifyCheckpoint` outside of `Initialize`.

Reply:
I moved this check codes in `NewRestoreController` right after `OpenCheckpointDB`